src/schnorr.rs

```diff
@@ -0,0 +1,107 @@
+    Zero,
+};
+use blake2::{
+    digest::{Update, VariableOutput},
+    VarBlake2b,
+};
+use std::ops::Neg;
+
+pub struct Schnorr<SC: SpongeConstants> {
+    pub sponge: ArithmeticSponge<PallasField, SC>,
+    pub network_id: NetworkId,
+}
+
+impl<SC: SpongeConstants> Signer for Schnorr<SC> {
+    fn sign<I: Input>(&mut self, kp: Keypair, input: I) -> Signature {
+        let k: PallasScalar = self.blinding_hash(&kp, input);
+        let r: PallasPoint = PallasPoint::prime_subgroup_generator().mul(k).into_affine();
+        let k: PallasScalar = if r.y.into_repr().is_even() { k } else { -k };
+
+        let e: PallasScalar = self.message_hash(&kp.pub_key, r.x, input);
+        let s: PallasScalar = k + e * kp.sec_key;
+
+        Signature::new(r.x, s)
+    }
+
+    fn verify<I: Input>(&mut self, sig: Signature, pub_key: PubKey, input: I) -> bool {
+        let ev: PallasScalar = self.message_hash(&pub_key, sig.rx, input);
+
+        let sv: PallasPoint = PallasPoint::prime_subgroup_generator()
+            .mul(sig.s)
+            .into_affine();
+        let rv: PallasPoint = sv + pub_key.mul(ev).neg().into_affine();
+
+        !rv.infinity && rv.y.into_repr().is_even() && rv.x == sig.rx
+    }
+}
+
+impl<SC: SpongeConstants> Schnorr<SC> {
+    pub fn new(sponge: ArithmeticSponge<PallasField, SC>, network_id: NetworkId) -> Self {
+        Schnorr::<SC> { sponge, network_id }
+    }
+
+    fn domain_bytes<I>(&self, input: I) -> Vec<u8>
+    where
+        I: Input,
+    {
+        let mut domain_string = input.domain_string(self.network_id);
+        // Domain prefixes have a max length of 20 and are padded with '*'
+        domain_string = &domain_string[..std::cmp::min(domain_string.len(), 20)];
+        let mut bytes = format!("{:*<20}", domain_string).as_bytes().to_vec();
+        bytes.resize(32, 0);
+
+        bytes
+    }
+
+    fn blinding_hash<I>(&mut self, kp: &Keypair, input: I) -> PallasScalar
+    where
+        I: Input,
+    {
+        let mut hasher = VarBlake2b::new(32).unwrap();
+
+        let mut roi: ROInput = input.to_roinput();
+        roi.append_field(kp.pub_key.x);
+        roi.append_field(kp.pub_key.y);
+        roi.append_scalar(kp.sec_key);
+        roi.append_bytes(vec![self.network_id.into()]);
+
+        hasher.update(roi.to_bytes());
+
+        let mut bytes = [0; 32];
+        hasher.finalize_variable(|out| bytes.copy_from_slice(out));
+        bytes[bytes.len() - 1] &= 0b0011_1111; // drop top two bits
+
+        PallasScalar::from_random_bytes(&bytes[..]).expect("failed to create scalar from bytes")
+    }
+
+    fn message_hash<I>(&mut self, pub_key: &PubKey, rx: PallasField, input: I) -> PallasScalar
+    where
+        I: Input,
+    {
+        let mut roi: ROInput = input.to_roinput();
+        roi.append_field(pub_key.x);
+        roi.append_field(pub_key.y);
+        roi.append_field(rx);
+
+        // Set sponge initial state (explicitly init state so signer context can be reused)
```

Comment:
what is the benefit of reusing that struct instead of creating a new one when you sign/verify something?

Reply:
From the user's perspective it relieves them from having to constantly create new signer contexts all the time.  Under the hood it eliminates the overhead of creating the context, initialising the mds and poseidon parameters.  I think it is also creates a safer and less error prone API because we don't need extra state and logic to prevent the user from using it more than once.  Similarly, the user doesn't need extra logic to keep track of whether a context has been used, etc.

Reply:
> Under the hood it eliminates the overhead of creating the context, initialising the mds and initialising poseidon parameters

I see, in this case a further optimization could also be to cache the state after permuting the domain, and just reusing that (instead of starting from state 0 next time).

Also, I'm a bit confused with the `self.sponge.squeeze();` line and why it's there? I'm guessing we do this on the ocaml side, but this doesn't seem immediately intuitive. Usually, you pad the domain separation string with enough 0 so that the next operation permutes. I guess it's equivalent to just squeeze when you know that you haven't reached the rate (here there's only one field element instead of 3). Maybe add a comment to make that clear? It almost feels like this API should be part of `oracle` and not directly used here.

Reply:
In Mina the init state of the sponge is not just the hash prefix string bytes packed into a field element and then padded into the sponge state.  Rather it seems the field element is actually absorbed into the sponge and then the state is squeezed (squoze?) out.